src/core/geometry/qgsrectangle.h

```diff
@@ -370,13 +370,15 @@ class CORE_EXPORT QgsRectangle
      */
     bool intersects( const QgsRectangle &rect ) const SIP_HOLDGIL
     {
-      const double x1 = ( mXmin > rect.mXmin ? mXmin : rect.mXmin );
-      const double x2 = ( mXmax < rect.mXmax ? mXmax : rect.mXmax );
-      if ( x1 > x2 )
+      if ( isNull() || rect.isNull() )
         return false;
-      const double y1 = ( mYmin > rect.mYmin ? mYmin : rect.mYmin );
-      const double y2 = ( mYmax < rect.mYmax ? mYmax : rect.mYmax );
-      return y1 <= y2;
+      const double x1 = ( ( std::isgreater( mXmin, rect.mXmin ) || std::isnan( mXmin ) ) ? mXmin : rect.mXmin );
```

Comment:
as the if eval start with the left most expression, I would have like that: 

```suggestion
      const double x1 = ( ( std::isnan( mXmin ) || std::isgreater( mXmin, rect.mXmin ) ) ? mXmin : rect.mXmin );
```